### PR TITLE
Better Component Initialization + prototype events component written in Rust 

### PR DESCRIPTION
**Pull request by Others**
### Summary of this Pull Request (PR)
Introduce a better API for libraries that need to do preinitialization, make libposix use this API, and add  a prototype `events` component written in Rust.

### Intent for your PR

Choose one (Mandatory):

- [ ] This PR is for a code-review and is intended to get feedback, but not to be pulled yet.
- [x] This PR is mature, and ready to be integrated into the repo.

### Reviewers (Mandatory):
(Specify @<github.com username(s)> of the reviewers. Ex: @user1, @user2)
@gparmer 
@vnitu02 you might want to chime in on the libposix stuff
@phanikishoreg I'd appreciate your input on how I define/use the `clone` syscall

### Code Quality

As part of this pull request, I've considered the following:

[Style](https://github.com/gparmer/composite/raw/ppos/doc/style_guide/composite_coding_style.pdf):

- [x] Comments adhere to the Style Guide (SG)
- [x] Spacing adhere's to the SG
- [x] Naming adhere's to the SG
- [x] All other aspects of the SG are adhered to, or exceptions are justified in this pull request
- [x] I have run the auto formatter on my code before submitting this PR (see doc/auto_formatter.md for instructions)

[Code Craftsmanship](http://www2.seas.gwu.edu/~gparmer/posts/2016-03-07-code-craftsmanship.html):

- [x] I've made an attempt to remove all redundant code
- [x] I've considered ways in which my changes might impact existing code, and cleaned it up
- [x] I've formatted the code in an effort to make it easier to read (proper error handling, function use, etc...)
- [x] I've commented appropriately where code is tricky
- [x] I agree that there is no "throw-away" code, and that code in this PR is of high quality

### Testing

I've tested the code using the following test programs (provide list here):
- `unit_*` tests
- `events_prototype`
- `libposix`

### CODE REVIEW

**Comment by Others** [comment]
I think the comment formatting is wrong in a lot of this code. I'll fix it when I do my first batch of changes :)

File: src/components/lib/cos_component.c [hunk 1]
```diff
@@ -21,16 +27,80 @@ cos_init(void *arg)
...
+void __init_libc(char **envp, char *pn);
+
+void
+libc_init()
+{
+	/* The construction of this is:
+	 * evn1, env2, ..., NULL, auxv_n1, auxv_1, auxv_n2, auxv_2 ..., NULL
+	 * TODO: Figure out a way to set AT_HWCAP / AT_SYSINFO
+	 */
+	static char *envp[] = {
```
Comment (gparmer) [inline src/components/lib/cos_component.c]: Is this the minimal set?  This is all essentially wasted memory, and isn't doing anything useful (but it is the easiest way to work with libc).  Thus minimizing its impact seems reasonable.

File: src/components/lib/posix/posix.c [hunk 2]
```diff
@@ -123,56 +110,173 @@ cos_mmap(void *addr, size_t length, int prot, int flags, int fd, off_t offset)
...
+int
+cos_clone(int (*func)(void *), void *stack, int flags, void *arg, pid_t *ptid, void *tls, pid_t *ctid)
 {
+	if(!func) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	struct sl_thd * thd = sl_thd_alloc((cos_thd_fn_t) func, arg);
+	if(tls) {
```
Comment (gparmer) [inline src/components/lib/posix/posix.c]: spaces after `if`.  Did auto-formatting not catch this?

File: src/components/lib/posix/posix.c [hunk 3]
```diff
@@ -123,56 +110,173 @@ cos_mmap(void *addr, size_t length, int prot, int flags, int fd, off_t offset)
...
+	setup_thread_area(sl_thd_curr(), data);
+	return 0;
 }
 
-CCTOR static void
-posix_init(void)
+int
+cos_clone(int (*func)(void *), void *stack, int flags, void *arg, pid_t *ptid, void *tls, pid_t *ctid)
 {
+	if(!func) {
```
Comment (gparmer) [inline src/components/lib/posix/posix.c]: space after `if`

File: src/components/lib/sl/sl_lock.c [hunk 4]
```diff
@@ -34,3 +34,16 @@ sl_lock_timed_take(struct sl_lock *lock, microsec_t max_wait_time)
     sl_cs_exit();
     return result;
 }
+
+int sl_lock_try_take(struct sl_lock *lock) {
```
Comment (gparmer) [inline src/components/lib/sl/sl_lock.c]: :+1:

**Pull request merged**